File: algorithms/2_medium/B/01145_binary_tree_coloring_game/dfs.c
enum states {
    NOT_SEEN_X,
    LEFT_X,
    RIGHT_X,
};

struct res_tuple {
    int not_seen_x;
    int left_x;
    int right_x;
};

static void dfs(struct TreeNode* node, const int x, const enum states state, struct res_tuple* res)
{
    if (node == NULL) {
        return;
    }

    if (node->val != x) {
        switch (state) {
            case NOT_SEEN_X: 
                res->not_seen_x += 1;
                break;
            case LEFT_X:
                res->left_x += 1;
                break;
            case RIGHT_X:
                res->right_x += 1;
                break;
            default:
                break;
        }
        dfs(node->left, x, state, res);
        dfs(node->right, x, state, res);
    } else {
        dfs(node->left, x, LEFT_X, res);
        dfs(node->right, x, RIGHT_X, res);
    }
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool btreeGameWinningMove(struct TreeNode* root, int n, int x)
{
    struct res_tuple res = { .not_seen_x = 0, .left_x = 0, .right_x = 0};
    dfs(root, x, NOT_SEEN_X, &res);

    return res.not_seen_x > res.left_x + res.right_x ||
           res.left_x > res.not_seen_x + res.right_x ||
           res.right_x > res.not_seen_x + res.left_x;
}

File: algorithms/2_medium/B/01145_binary_tree_coloring_game/dfs.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

#[derive(Clone, Copy)]
enum States {
    NotSeenX,
    LeftX,
    RightX,
}

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    fn dfs(node_rc_opt: Option<&Rc<RefCell<TreeNode>>>, x: i32, state: States, res: &mut (i32, i32, i32)) {
        if let Some(node_rc) = node_rc_opt {
            let node = node_rc.borrow();
            if node.val != x {
                match state {
                    States::NotSeenX => res.0 += 1,
                    States::LeftX => res.1 += 1,
                    States::RightX => res.2 += 1,
                }
                Self::dfs(node.left.as_ref(), x, state, res);
                Self::dfs(node.right.as_ref(), x, state, res);
            } else {
                Self::dfs(node.left.as_ref(), x, States::LeftX, res);
                Self::dfs(node.right.as_ref(), x, States::RightX, res);
            }
        }
    }

    pub fn btree_game_winning_move(root: Option<Rc<RefCell<TreeNode>>>, n: i32, x: i32) -> bool {
        let mut res = (0_i32, 0_i32, 0_i32);
        Self::dfs(root.as_ref(), x, States::NotSeenX, &mut res);

        res.0 > res.1 + res.2 ||
        res.1 > res.0 + res.2 ||
        res.2 > res.0 + res.1
    }
}
